cmd/charm-ingest/whitelist.go

```diff
@@ -0,0 +1,65 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the GPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/juju/charmstore-client/internal/ingest"
+	"gopkg.in/juju/charmrepo.v4/csclient/params"
+)
+
+var validChannels = params.ValidChannels
+
+// parseWhitelist takes an io.Reader for a whitelist and returns []ingest.WhitelistEntity
+// or an error if unable to parse
+func parseWhitelist(r io.Reader) ([]ingest.WhitelistEntity, error) {
+	var entities []ingest.WhitelistEntity
+	fileScanner := bufio.NewScanner(r)
+	lineNum := 0
+
+	for fileScanner.Scan() {
+		lineNum++
+
+		entity, err := parseLine(fileScanner.Text(), lineNum)
+		if err != nil {
+			return nil, err
+		}
+
+		if entity.EntityId == "" {
+			// Empty line, ignore
+			continue
+		}
+		entities = append(entities, entity)
+	}
+
+	return entities, nil
+}
+
+// parseLine takes a string line of a whitelist file with its corresponding line number,
+// and returns the ingest.WhitelistEntity for that line, or an error
+func parseLine(line string, lineNum int) (ingest.WhitelistEntity, error) {
```

Comment:
Passing a line number in here just so that it can be used in a potential error message seems backwards to me. It would be better to annotate the error with the line number (and probably file) later.

Something like:
```
type parseError struct {
	Filename string
	Line int
	Err error
}

func (e *parseError) Error() string {
	return fmt.Sprintf("%s:%d: %v", e.FileName, e.Line, e.Err)
}

func parseWhitelist(filename string, r io.Reader) ([]ingest.WhitelistEntity, error) {
	var entities []ingest.WhitelistEntity
	fileScanner := bufio.NewScanner(r)
	lineNum := 0
	for fileScanner.Scan() {
		lineNum++
		entity, err := parseLine(fileScanner.Text())
		if err != nil {
			return nil, &parseError{
				Filename: filename,
				Line: lineNum,
				Err: err,
			}
		}
		if entity.EntityId == "" {
			// Empty line, ignore
			continue
		}
		entities = append(entities, entity)
	}
	return entities, nil
}
```